components/eamxx/src/physics/gw/impl/gw_gwd_compute_tendencies_from_stress_divergence_impl.hpp

```diff
@@ -20,27 +27,150 @@ void Functions<S,D>::gwd_compute_tendencies_from_stress_divergence(
   const Int& pgwv,
   const Int& ngwv,
   const bool& do_taper,
-  const Spack& dt,
-  const Spack& effgw,
+  const Real& dt,
+  const Real& effgw,
   const uview_1d<const Int>& tend_level,
-  const uview_1d<const Spack>& lat,
-  const uview_1d<const Spack>& dpm,
-  const uview_1d<const Spack>& rdpm,
-  const uview_1d<const Spack>& c,
-  const uview_1d<const Spack>& ubm,
-  const uview_1d<const Spack>& t,
-  const uview_1d<const Spack>& nm,
-  const uview_1d<const Spack>& xv,
-  const uview_1d<const Spack>& yv,
+  const uview_1d<const Real>& lat,
+  const uview_2d<const Real>& dpm,
+  const uview_2d<const Real>& rdpm,
+  const uview_2d<const Real>& c,
+  const uview_2d<const Real>& ubm,
+  const uview_2d<const Real>& t,
+  const uview_2d<const Real>& nm,
+  const uview_1d<const Real>& xv,
+  const uview_1d<const Real>& yv,
   // Inputs/Outputs
-  const uview_1d<Spack>& tau,
+  const uview_3d<Real>& tau,
   // Outputs
-  const uview_1d<Spack>& gwut,
-  const uview_1d<Spack>& utgw,
-  const uview_1d<Spack>& vtgw)
+  const uview_3d<Real>& gwut,
+  const uview_2d<Real>& utgw,
+  const uview_2d<Real>& vtgw)
 {
-  // TODO
-  // Note, argument types may need tweaking. Generator is not always able to tell what needs to be packed
+  // Temporary ubar tendencies (overall, at wave l, and at saturation).
+  // These should probably come from workspace manager
+  view_2d<Real> ubt("ubt", ncol, s_common_init.pver);
+  view_1d<Real> ubtl("ubtl", ncol), ubtlsat("ubtlsat", ncol);
+
+  // Polar taper.
+  view_1d<Real> ptaper("ptaper", ncol);
+
+  if (do_taper) { // taper CM only
+    for (int i = 0; i < ncol; ++i) {
+      ptaper(i) = std::cos(lat(i));
+    }
+  }
+  else { //  do not taper other cases
+    Kokkos::deep_copy(ptaper, 1.);
+  }
+
+  // Force tau at the top of the model to zero, if requested.
+  if (s_common_init.tau_0_ubc) {
+    for (size_t i = 0; i < tau.extent(0); ++i) {
+      for (size_t j = 0; j < tau.extent(1); ++j) {
+        tau(i,j,0) = 0.;
+      }
+    }
+  }
+
+  // Find max
+  int maxval = -999999;
+  for (int i = 0; i < ncol; ++i) {
+    if (tend_level(i) > maxval) maxval = tend_level(i);
+  }
+
+  // Loop over levels from top to bottom
+  for (int k = s_common_init.ktop; k <= maxval; ++k) {
+    //  Accumulate the mean wind tendency over wavenumber.
+    for (int i = 0; i < ncol; ++i) {
+      ubt(i,k) = 0.;
+    }
+
+    for (int l = -ngwv; l <= ngwv; ++l) { // loop over wave
+      int nl_idx = l + ngwv; // 0-based idx for -ngwv:ngwv arrays
+      int pl_idx = nl_idx + (pgwv - ngwv); // 0-based idx -pgwv:pgwv arrays
+
+      // Determine the wind tendency, including excess stress carried down
+      // from above.
+      for (int i = 0; i < ncol; ++i) {
```

Comment:
I suspect you'll want to put the `k` loop inside multiple `ncol` loops and parallelize as in P3 and SHOC: top-level TeamPolicy parallel_for over columns, then team parallelism over levels.

Reply:
I agree. The question is, will the top-level col kernel already be launched before this function is called? That's what we did in p3.

Reply:
Yes, Luca mentioned that. I think we are all in agreement then, thx!